common/src/main/java/io/pravega/common/concurrent/ServiceHelpers.java

```diff
@@ -71,32 +72,49 @@
      * @param executor           An Executor to use for callback invocations.
      */
     public static void onStop(Service service, Runnable terminatedCallback, Consumer<Throwable> failureCallback, Executor executor) {
+        ShutdownListener listener = new ShutdownListener(terminatedCallback, failureCallback);
+        service.addListener(listener, executor);
+
+        // addListener() will not invoke the callbacks if the service is already in a terminal state. As such, we need to
```

Comment:
`addListener()` will attach a Listener - nothing more. If the service is already in a TERMINATED or FAILED state, the Listener will never get invoked. That was already handled before. However, it is possible that the service was not in one of these terminal states when we did the check, however, between that check and the time we register the Listener it would have transitioned to such a state - in that case the callback would have never gotten invoked.

This change moves the one-time check after the registration and also adds the code in the Listener to make sure that the callbacks are only invoked once - it is possible that the Service transitions to a terminal state after registering the Listener, but before the check, at which point the callbacks would be invoked twice.